File: LeetCode/Java/src/solutions/p0133_CloneGraph/P0133.java
package solutions.p0133_CloneGraph;

import auxiliary.Node;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class P0133 {
    public Node cloneGraph(Node node) {
        if (node == null) {
            return null;
        }

        Map<Node, Node> ans = new HashMap<>();
        ans.put(node, new Node(node.val));

        Stack<Node> stack = new Stack<>();
        stack.add(node);
        while (!stack.empty()) {
            Node curr = stack.pop();

            for (Node n : curr.neighbors) {
                if (ans.get(n) == null) {
                    ans.put(n, new Node(n.val));
                    stack.add(n);
                }
                ans.get(curr).neighbors.add(ans.get(n));
            }
        }
        return ans.get(node);
    }
}

File: LeetCode/Python/solutions/p0133.py
from Python.auxiliary.Node import Node


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':

        if not node:
            return None

        if not node.neighbors:
            return Node(node.val)

        ans = {node: Node(node.val)}
        q = [node]
        while q:
            curr = q.pop()

            for n in curr.neighbors:
                if n not in ans:
                    ans[n] = Node(n.val)
                    q.append(n)
                ans[curr].neighbors.append(ans[n])

        return ans[node]
